fe/parameterize_batch.py

```diff
@@ -0,0 +1,381 @@
+
+def com(conf):
+    return np.sum(conf, axis=0)/conf.shape[0]
+
+def recenter(conf, true_com, scale_factor=1):
+    mol_com = np.sum(conf, axis=0)/conf.shape[0]
+    centered = conf - mol_com  # centered to origin
+    return true_com + centered/scale_factor 
+
+from hilbertcurve.hilbertcurve import HilbertCurve
+
+def hilbert_sort(conf):
+    hc = HilbertCurve(16, 3)
+    int_confs = (conf*1000).astype(np.int64)+10000
+    dists = []
+    for xyz in int_confs.tolist():
+        dist = hc.distance_from_coordinates(xyz)
+        dists.append(dist)
+    perm = np.argsort(dists)
+    return perm
+
+class PDBWriter():
+
+    def __init__(self, pdb_str, out_filepath):
+        self.pdb_str = pdb_str
+        self.out_filepath = out_filepath
+        self.outfile = None
+        self.n_frames = 25
+
+    def write_header(self):
+        """
+        Confusingly this initializes writer as well because 
+        """
+        outfile = open(self.out_filepath, 'w')
+        self.outfile = outfile
+        cpdb = app.PDBFile(self.pdb_str)
+        PDBFile.writeHeader(cpdb.topology, self.outfile)
+        self.topology = cpdb.topology
+        self.frame_idx = 0
+
+    def write(self, x):
+        if self.outfile is None:
+            raise ValueError("remember to call write_header first")
+        self.frame_idx += 1
+        PDBFile.writeModel(self.topology, x, self.outfile, self.frame_idx)
+
+    def close(self):
+        PDBFile.writeFooter(self.topology, self.outfile)
+        self.outfile.flush()
+
+import jax.numpy as jnp
+
+def error_fn(all_du_dls, T, schedule, true_dG):
+    fwd = all_du_dls[:, :T//2]
+    fwd_sched = schedule[:T//2]
+    bkwd = all_du_dls[:, T//2:]
+    bkwd_sched = schedule[T//2:]
+    dG_fwd = math_utils.trapz(fwd, fwd_sched) # integral from inf to 0
+    dG_bkwd = math_utils.trapz(bkwd, bkwd_sched) # integral from 0 to inf
+    # dG_fwd and dG_bkwd have the same sign, so we need to flip dG_bkwd so the
+    # direction of integral is the same (requirement for pymbar.BAR)
+    dG_bkwd = -dG_bkwd # this is needed for BAR to be correct
+
+    # this is in kJ/mol, inputs to BAR needs to be in 1/kT.
+    kT = 2.479
+    dG_fwd /= kT
+    dG_bkwd /= kT
+    pred_dG = loss.mybar(jnp.stack([dG_fwd, dG_bkwd]))
+    pred_dG *= kT
+
+    return jnp.abs(pred_dG - true_dG)
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(description='Quick Test')
+    parser.add_argument('--frames_dir', type=str, required=True)
+    parser.add_argument('--precision', type=str, required=True)    
+    parser.add_argument('--complex_pdb', type=str, required=True)
+    parser.add_argument('--ligand_sdf_file', type=str, nargs='*')
+    parser.add_argument('--num_gpus', type=int, required=True)
+    parser.add_argument('--jobs_per_gpu', type=int, required=True)
+    parser.add_argument('--num_conformers', type=int, required=True)
+    parser.add_argument('--true_dG_file', type=str, nargs='*')
```

Comment:
I would add a docstring clarifying the sign of the dG values (either insertion or deletion) and if they're in kJ/mol or kcal/mol. This (along with ligand_sdf_file) should be required.